### PR TITLE
Enable resolution of queries with negative literals in probabilistic resolution + fix #574 & #575

### LINKED ISSUE #574: Dead code?
https://github.com/NeuroLang/NeuroLang/blob/0984fd887e99960ad6453709c451e986d6780969/neurolang/relational_algebra_provenance.py#L300

Maybe I'm wrong but my understanding is that this line does not do anything as `symmetric_difference` returns a new set

### PR DESCRIPTION
**Pull request by tgy**

### CODE REVIEW

File: neurolang/probabilistic/stratification.py [hunk 1]
```diff
@@ -129,9 +143,9 @@ def _get_program_deterministic_symbols(program):
 
 def _get_rule_idb_type(rule, grpd_symbs, wlq_symbs):
     dep_symbs = set(
-        pred.functor
-        for pred in extract_logic_atoms(rule.antecedent)
-        if isinstance(pred.functor, Symbol)
```
Comment (demianw) [inline neurolang/probabilistic/stratification.py:134]: Atoms already extract the atom without the negation. I think this part worked before

File: neurolang/probabilistic/stratification.py [hunk 2]
```diff
@@ -101,12 +100,27 @@ def stratify_program(query, program):
...
             count = len(idb)
+    _check_no_negation_in_probabilistic_rules(grpd_idbs["probabilistic"])
     return {
         idb_type: Union(tuple(idb_rules))
         for idb_type, idb_rules in grpd_idbs.items()
+        if idb_rules
     }
 
 
+def _check_no_negation_in_probabilistic_rules(prob_rules):
```
Comment (demianw) [inline neurolang/probabilistic/stratification.py:111]: maybe modify this check according to what we discussed? We allow negated atoms which are on the EDB.

**Review by demianw — CHANGES_REQUESTED** [review]
Good job. Small comments

File: neurolang/probabilistic/cplogic/testing.py [hunk 3]
```diff
@@ -50,9 +50,12 @@ def get_named_relation_tuples(relation):
 def eq_prov_relations(pas1, pas2):
     assert isinstance(pas1, ProvenanceAlgebraSet)
     assert isinstance(pas2, ProvenanceAlgebraSet)
+    pas1_sorted_np_cols = sorted(pas1.non_provenance_columns)
+    pas2_sorted_np_cols = sorted(pas2.non_provenance_columns)
+    assert pas1_sorted_np_cols == pas2_sorted_np_cols
     assert (
         pas1.value.projection(*pas1.non_provenance_columns).to_unnamed()
-        == pas2.value.projection(*pas2.non_provenance_columns).to_unnamed()
+        == pas2.value.projection(*pas1.non_provenance_columns).to_unnamed()
```
Comment (tgy) [inline neurolang/probabilistic/cplogic/testing.py:58]: small fix to make sure the provenance sets are projected on similarly ordered columns

File: neurolang/probabilistic/cplogic/tests/test_cplogic_program.py [hunk 4]
```diff
@@ -97,6 +103,33 @@ def test_cplogic_program():
...
+class CPLogicNegationProgram(
+    CPLogicMixin,
+    DatalogProgramNegationMixin,
+    DatalogProgram,
+    ExpressionWalker,
+):
+    pass
+
+
+def test_negated_predicate_in_antecedent():
```
Comment (tgy) [inline neurolang/probabilistic/cplogic/tests/test_cplogic_program.py:115]: this test is here just to make sure that a rule with negation is correctly added to the program

**Review by demianw — APPROVED** [review]

**Pull request merged**